src/python/WMComponent/RucioInjector/RucioInjectorPoller.py

```diff
@@ -0,0 +1,426 @@
+                        listLfns.append(fileInfo['lfn'])
+                        injectData.append(dict(name=fileInfo['lfn'], scope=self.scope,
+                                               bytes=fileInfo['size'], state="A",
+                                               adler32=fileInfo['checksum']['adler32']))
+
+                    if self.rucio.createReplicas(rse=rseName, files=injectData, block=block):
+                        logging.info("Successfully inserted %d files on block %s", len(listLfns), block)
+                        self._updateLFNState(listLfns)
+        return
+
+    def _updateLFNState(self, listLfns, recovery=False):
+        """
+        Given a list of LFNs, update their state in dbsbuffer table.
+        :param listLfns: list of LFNs
+        :param recovery: True if we are recovering previously injected files
+        :return: nothing
+        """
+        if not listLfns:
+            return
+        try:
+            self.setStatus.execute(listLfns, 1)
+        except Exception as ex:
+            # save it to try to inject them again in the next cycle
+            self.filesToRecover.extend(listLfns)
+            if 'Deadlock found' in str(ex) or 'deadlock detected' in str(ex):
+                logging.error("Deadlock during file status update. Retrying again in the next cycle.")
+                self.filesToRecover.extend(listLfns)
+            else:
+                msg = "Failed to update file status in the database, reason: %s" % str(ex)
+                logging.error(msg)
+                raise RucioInjectorException(msg)
+        else:
+            if recovery:
+                self.filesToRecover = []
+
+    def closeBlocks(self):
+        """
+        Close any blocks that have been migrated to global DBS
+        """
+        logging.info("Starting closeBlocks method")
+
+        # in short, dbsbuffer_file.in_phedex = 1 AND dbsbuffer_block.status = 'InDBS'
+        migratedBlocks = self.getMigrated.execute()
+        ### FIXME the data format returned by this DAO
+        for location in migratedBlocks:
+            for container in migratedBlocks[location]:
+                for block in migratedBlocks[location][container]:
+                    if self.rucio.closeBlockContainer(block):
+                        self.setBlockClosed.execute(block)
+                    else:
+                        logging.error("Failed to close block: %s. Will retry again later.", block)
+
+    def deleteBlocks(self):
+        """
+        _deleteBlocks_
+        Find deletable blocks, then decide if to delete based on:
+        Is there an active subscription for dataset or block ?
+          If yes => set deleted=2
+          If no => next check
+        Has transfer to all destinations finished ?
+          If yes => request block deletion, approve request, set deleted=1
+          If no => do nothing (check again next cycle)
+        """
+        raise NotImplementedError
```

Comment:
Here you should be able to delete the rules you made on the blocks if you can find them or kept track of them. If a another rule was made to move the container to another site and it hasn't happened yet, Rucio is smart enough to know this and will not clean up the replicas

Reply:
In other words I don't think you even need to find the deletable blocks.

Reply:
Yes, I will address the rule tracking in a different issue. For now, I'm leaving just a placeholder for rules deletion. I have to review what triggers the T0 rule deletion as well. I believe what we will need is to just check whether the DID has any other rule on it, no?